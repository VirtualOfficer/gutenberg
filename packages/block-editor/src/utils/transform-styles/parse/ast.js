/**
 * External dependencies
 */
import CSSwhat from 'css-what';

function cssSelectorAstInArray( objNeedle, arr ) {
	return arr.some(
		( obj ) => JSON.stringify( obj[ 0 ] ) === JSON.stringify( objNeedle )
	);
}

function createWrapperSelectorAst( wrapperSelectorCss ) {
	const wrapperSelectorAst = CSSwhat.parse( wrapperSelectorCss ); // (double-nested)
	wrapperSelectorAst[ 0 ].push( {
		type: 'descendant',
	} );
	return wrapperSelectorAst;
}

function recurseCssRules( cssRule, callback ) {
	if ( ! cssRule.cssRules ) {
		return;
	}
	for ( const subCssRule of cssRule.cssRules ) {
		callback( subCssRule );
		recurseCssRules( subCssRule, callback );
	}
}

export { cssSelectorAstInArray, createWrapperSelectorAst, recurseCssRules };
