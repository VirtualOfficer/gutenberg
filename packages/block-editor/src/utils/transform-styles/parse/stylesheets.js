function createStyleElem( css, hidden = true ) {
	const styleElem = document.createElement( 'style' );

	if ( hidden ) {
		styleElem.media = 'none';
	}

	document.head.appendChild( styleElem );
	styleElem.appendChild( document.createTextNode( css ) );

	if ( hidden ) {
		styleElem.sheet.disabled = true;
	}

	return styleElem;
}

function textFromStyleSheet( styleSheet ) {
	const cssText = Array.from( styleSheet.cssRules ).reduce(
		( prev, cssRule ) => {
			return prev + cssRule.cssText;
		},
		''
	);
	return cssText;
}

export { createStyleElem, textFromStyleSheet };
