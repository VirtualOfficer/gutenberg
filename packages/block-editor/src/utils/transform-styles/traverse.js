/**
 * External dependencies
 */
import csstree from 'css-tree';

function traverseCSS( css, callback ) {
	try {
		const ast = csstree.parse( css );
		csstree.walk( ast, callback );
		return csstree.generate( ast );
	} catch ( err ) {
		// eslint-disable-next-line no-console
		console.warn( 'Error while traversing the CSS: ' + err );

		return null;
	}
}

export default traverseCSS;
