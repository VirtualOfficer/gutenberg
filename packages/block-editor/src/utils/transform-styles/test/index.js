/**
 * Internal dependencies
 */
import traverse from '../traverse';
import parse from '../ast/parse';

describe( 'CSS traverse', () => {
	it( 'Should traverse the CSS', () => {
		const input = `h1 { color: red; }`;
		const output = traverse( input, ( node ) => {
			if ( node.type === 'rule' ) {
				return {
					...node,
					selectors: node.selectors.map(
						( selector ) => 'namespace ' + selector
					),
				};
			}

			return node;
		} );

		expect( output ).toMatchSnapshot();
	} );
} );

describe( 'CSS parse', () => {
	it( 'Should handle closing parenthesis within url', () => {
		const input = `.a { b: url(")"); d: url(";a"); }`;
		const output = parse( input );

		expect( output ).toMatchSnapshot();
	} );

	it( 'Should handle urls without quotes', () => {
		const input = `.a {
			background-image: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAQAAAAEBAMAAABb34NNAAAAFVBMVEXyrmHqcBDtdQ3pcRHcbRkAAAD/uSmYi/+cAAAABnRSTlORbFE2DwAwkbURAAAAEUlEQVQI12MIEmEwSwAiIAMAC/AB+fq6WIgAAAAASUVORK5CYII=);
			background-repeat: no-repeat
		}`;
		const output = parse( input );

		expect( output ).toMatchSnapshot();
	} );
} );
