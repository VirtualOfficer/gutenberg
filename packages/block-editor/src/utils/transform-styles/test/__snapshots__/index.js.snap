// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CSS parse Should handle closing parenthesis within url 1`] = `
Object {
  "stylesheet": Object {
    "parsingErrors": Array [],
    "rules": Array [
      Object {
        "declarations": Array [
          Object {
            "position": Position {
              "end": Object {
                "column": 17,
                "line": 1,
              },
              "source": undefined,
              "start": Object {
                "column": 6,
                "line": 1,
              },
            },
            "property": "b",
            "type": "declaration",
            "value": "url(\\")\\")",
          },
          Object {
            "position": Position {
              "end": Object {
                "column": 31,
                "line": 1,
              },
              "source": undefined,
              "start": Object {
                "column": 19,
                "line": 1,
              },
            },
            "property": "d",
            "type": "declaration",
            "value": "url(\\";a\\")",
          },
        ],
        "position": Position {
          "end": Object {
            "column": 34,
            "line": 1,
          },
          "source": undefined,
          "start": Object {
            "column": 1,
            "line": 1,
          },
        },
        "selectors": Array [
          ".a",
        ],
        "type": "rule",
      },
    ],
    "source": undefined,
  },
  "type": "stylesheet",
}
`;

exports[`CSS parse Should handle urls without quotes 1`] = `
Object {
  "stylesheet": Object {
    "parsingErrors": Array [],
    "rules": Array [
      Object {
        "declarations": Array [
          Object {
            "position": Position {
              "end": Object {
                "column": 217,
                "line": 2,
              },
              "source": undefined,
              "start": Object {
                "column": 4,
                "line": 2,
              },
            },
            "property": "background-image",
            "type": "declaration",
            "value": "url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAQAAAAEBAMAAABb34NNAAAAFVBMVEXyrmHqcBDtdQ3pcRHcbRkAAAD/uSmYi/+cAAAABnRSTlORbFE2DwAwkbURAAAAEUlEQVQI12MIEmEwSwAiIAMAC/AB+fq6WIgAAAAASUVORK5CYII=)",
          },
          Object {
            "position": Position {
              "end": Object {
                "column": 3,
                "line": 4,
              },
              "source": undefined,
              "start": Object {
                "column": 4,
                "line": 3,
              },
            },
            "property": "background-repeat",
            "type": "declaration",
            "value": "no-repeat",
          },
        ],
        "position": Position {
          "end": Object {
            "column": 4,
            "line": 4,
          },
          "source": undefined,
          "start": Object {
            "column": 1,
            "line": 1,
          },
        },
        "selectors": Array [
          ".a",
        ],
        "type": "rule",
      },
    ],
    "source": undefined,
  },
  "type": "stylesheet",
}
`;

exports[`CSS traverse Should traverse the CSS 1`] = `
"namespace h1 {
color: red;
}"
`;
